sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala

```diff
@@ -80,28 +105,57 @@ abstract class QueryPlan[PlanType <: QueryPlan[PlanType]]
    * transformExpressionsDown or transformExpressionsUp should be used.
    *
    * @param rule the rule to be applied to every expression in this operator.
+   * @param cond  a Lambda expression to prune tree traversals. If `cond.apply` returns false
+   *              on an expression T, skips processing T and its subtree; otherwise, processes
+   *              T and its subtree recursively.
+   * @param ruleId is a unique Id for `rule` to prune unnecessary tree traversals. When it is
+   *        RuleId.UnknownId, no pruning happens. Otherwise, if `rule`(with id `ruleId`) has been
+   *        marked as in effective on an expression T, skips processing T and its subtree. Do not
+   *        pass it if the rule is not purely functional and reads a varying initial state for
+   *        different invocations.
    */
-  def transformExpressions(rule: PartialFunction[Expression, Expression]): this.type = {
-    transformExpressionsDown(rule)
+  def transformExpressions(rule: PartialFunction[Expression, Expression],
```

Comment:
Can we make this signature a bit less clunky? I would prefer something like this:
```scala
def transformExpressions(ruleId: Int, cond: TreePatternBits)(rule: PartialFunction[Expression, Expression): this.type = {
  ...
}

// Use it
plan.transformExpressions(MyRule.id, _.contains(SOME_PATTERN) {
  case plan => plan
}
```

Reply:
I remember @dbaliafroozeh and I discussed that previously. The signature and its variation (e.g., (rule)(ruleId, cond)) requires us to change all call sites of transformXxx in a single PR because existing code like "plan transform {...}" wouldn't compile?

Reply:
We can create an overload for the current code right?
```scala
def transformExpressions(rule: PartialFunction[Expression, Expression): this.type = {
  transformExpressions(RuleIdCollection.UnknownId, AlwaysProcess.fn)(rule)
}

Reply:
yeah, that works. I'll go this way.

Reply:
During the code transition period, recursion stack depth would be doubled, but I think it's probably fine.

Reply:
It would increase the stack depth by 1 right? The other transformDown function should not call into this one. If you think it is an issue, you could annotate this with `@inline`.

Reply:
>> It would increase the stack depth by 1 right

You're right. For recursive calls inside `transformDown`, we should always call the new overload with `ruleId` and `cond`. I'll add an overload for each transform function.

Reply:
I tried to overload transform functions, but ran into the following issue:

- the overload does work with Scala 2.13, but doesn't work with Scala 2.11 nor the default Scala version Spark build uses.  Basically, the overload fails existing call sites like `e transform {....}`. To repro what happens, we can do the following in a Scala console:
        scala>  object X { def f(i: Int) = i+1 ; def f(g: String=> String) = g("a") }
                      defined object X

        scala>  X.f {case s => s + "b"}
                     <console>:13: error: missing parameter type for expanded function
                     The argument types of an anonymous function must be fully known. (SLS 8.5)
                     Expected type was: ?
                      X.f {case s => s + "b"}

- also, I found that style guide mentioned to not use multiple parameter lists: https://github.com/databricks/scala-style-guide#multiple-parameter-lists

Let me know if you have a better thought.

Reply:
Done.  Used a different function name, i.e., transformWithPruning, per our offline discussion.